### PR TITLE
Imlementation of AtUri proposal

### LINKED ISSUE #36: Proposal: Utilizing AtUri in place of Strings
Working Branch: [afbase:afbase/using_aturi](https://github.com/afbase/rsky/tree/afbase/using_aturi)

I'm happy to get feedback and make a PR if you'd like.

# Proposal: Implement TryFrom for AtUri string conversions

## Overview
This PR implements the [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) trait for `AtUri` to handle string-to-AtUri conversions. This addresses multiple TODOs throughout the codebase where string manipulation was being used instead of proper AtUri types.

## Changes

Files modified:
- src/apis/com/atproto/admin/update_subject_status.rs
- src/apis/com/atproto/repo/apply_writes.rs
- src/apis/com/atproto/repo/create_record.rs
- src/apis/com/atproto/repo/delete_record.rs
- src/apis/com/atproto/repo/get_record.rs
- src/apis/com/atproto/repo/list_records.rs
- src/apis/com/atproto/repo/put_record.rs
- src/repo/mod.rs
- src/repo/record/mod.rs
- src/repo/types.rs
- src/syntax/aturi.rs

## Implementation Details

The core change is implementing TryFrom for AtUri with three variants:
```rust
impl TryFrom<&str> for AtUri
impl TryFrom<String> for AtUri  
impl TryFrom<&String> for AtUri
```

This allows for ergonomic conversion of strings to AtUri types using the `try_into()` method, replacing manual string manipulation throughout the codebase, e.g.:

```rust
// Before
let uri_without_prefix = uri.replace("at://", "");
let parts = uri_without_prefix.split("/").collect::<Vec<&str>>();

// After
let at_uri: AtUri = uri.try_into()?;
```

## Review Notes

1. Please review the `get_backlink_conflicts` function in `rsky-pds/src/repo/record/mod.rs` - specifically the use of [`flat_map`](https://github.com/afbase/rsky/blob/cf614eb84a036db31f7bee52072ea5075bc2feb1/rsky-pds/src/repo/record/mod.rs#L318-L323). I chose `flat_map` since we're dealing with `Option<AtUri>`, but I'm not 100% sure if this is the optimal approach vs using `map`. Would appreciate input here.

2. **TODO**: AtUri validation still needs to be implemented. I plan to add validation logic that mirrors the TypeScript implementation found at: https://github.com/bluesky-social/atproto/blob/main/packages/syntax/src/aturi_validation.ts - I don't intend on implementing the validation in the branch unless you'd prefer it that way.  It's already a fairly large PR ready to go.

## Testing
I added some test cases for the TryFrom implementations, including:
- Basic URI conversions
- URIs with query parameters
- URIs with hash fragments
- Invalid URI handling
- Edge cases

Let me know if any additional test cases should be considered.

### PR DESCRIPTION
**Pull request by afbase**
closes #36 

## Overview
This PR implements the [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) trait for `AtUri` to handle string-to-AtUri conversions. This addresses multiple TODOs throughout the codebase where string manipulation was being used instead of proper AtUri types.

## Changes

Files modified:
- src/apis/com/atproto/admin/update_subject_status.rs
- src/apis/com/atproto/repo/apply_writes.rs
- src/apis/com/atproto/repo/create_record.rs
- src/apis/com/atproto/repo/delete_record.rs
- src/apis/com/atproto/repo/get_record.rs
- src/apis/com/atproto/repo/list_records.rs
- src/apis/com/atproto/repo/put_record.rs
- src/repo/mod.rs
- src/repo/record/mod.rs
- src/repo/types.rs
- src/syntax/aturi.rs

## Implementation Details

The core change is implementing TryFrom for AtUri with three variants:
```rust
impl TryFrom<&str> for AtUri
impl TryFrom<String> for AtUri  
impl TryFrom<&String> for AtUri
```

This allows for ergonomic conversion of strings to AtUri types using the `try_into()` method, replacing manual string manipulation throughout the codebase, e.g.:

```rust
// Before
let uri_without_prefix = uri.replace("at://", "");
let parts = uri_without_prefix.split("/").collect::<Vec<&str>>();

// After
let at_uri: AtUri = uri.try_into()?;
```

## Review Notes

1. Please review the `get_backlink_conflicts` function in `rsky-pds/src/repo/record/mod.rs` - specifically the use of [`flat_map`](https://github.com/afbase/rsky/blob/cf614eb84a036db31f7bee52072ea5075bc2feb1/rsky-pds/src/repo/record/mod.rs#L318-L323). I chose `flat_map` since we're dealing with `Option<AtUri>`, but I'm not 100% sure if this is the optimal approach vs using `map`. Would appreciate input here.

2. **TODO**: AtUri validation still needs to be implemented. I plan to add validation logic that mirrors the TypeScript implementation found at: https://github.com/bluesky-social/atproto/blob/main/packages/syntax/src/aturi_validation.ts - I don't intend on implementing the validation in the branch unless you'd prefer it that way.  It's already a fairly large PR ready to go.

### CODE REVIEW

File: rsky-pds/src/repo/record/mod.rs [hunk 1]
```diff
@@ -321,110 +315,105 @@ impl RecordReader {
...
             .into_iter()
             .flatten()
-            .map(|record| {
-                format!(
-                    "at://{0}/{1}/{2}",
-                    env::var("PDS_HOSTNAME").unwrap_or("localhost".to_owned()),
-                    collection,
-                    record.rkey
-                )
+            .flat_map(|record| {
```
Comment (rudyfraser) [inline rsky-pds/src/repo/record/mod.rs:318]: While I'm sure the performance is the same, I think the following would be clearer:
```
Ok(conflicts)
    .into_iter()
    .flatten()
    .filter_map(|record| {
        AtUri::make(
            env::var("PDS_HOSTNAME").unwrap_or("localhost".to_owned()),
            Some(String::from(uri.get_collection())),
            Some(record.rkey),
        )
        .ok()
    })
    .collect::<Vec<AtUri>>()
```

That approach more directly expresses the intent:
1. 	Generate an Option<AtUri> for each record.
2. 	Skip (filter out) any that are None.
3. 	Collect the remaining AtUris into a Vec.

Let me know what you think.

**Review by rudyfraser — CHANGES_REQUESTED** [review]
Left a comment about updating the get_backlink_conflicts function but everything else LGTM, thank you

File: rsky-pds/src/repo/record/mod.rs (hunk 1 above)
Comment (afbase) [reply rsky-pds/src/repo/record/mod.rs]: LGTM

resolved with 
```rust
Ok(conflicts
        .into_iter()
        .flatten()
        .filter_map(|record| {
            AtUri::make(
                env::var("PDS_HOSTNAME").unwrap_or("localhost".to_owned()),
                Some(String::from(uri.get_collection())),
                Some(record.rkey),
            )
            .ok()
        })
        .collect::<Vec<AtUri>>())
```
